node/api.go

```diff
@@ -61,83 +58,55 @@ func (api *PrivateAdminAPI) AddPeer(url string) (bool, error) {
-	var offeredAPIs []rpc.API
-	if len(apis) > 0 {
-		namespaces := set.New()
-		for _, a := range strings.Split(apis, ",") {
-			namespaces.Add(strings.TrimSpace(a))
-		}
-		for _, api := range api.node.APIs() {
-			if namespaces.Has(api.Namespace) {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	} else { // use by default all public API's
-		for _, api := range api.node.APIs() {
-			if api.Public {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	}
+func (api *PrivateAdminAPI) StartRPC(host string, port int, cors string, apis string) (bool, error) {
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
 
-	if address == "" {
-		address = "127.0.0.1"
+	if api.node.httpHandler != nil {
+		return false, fmt.Errorf("HTTP RPC already running on %s", api.node.httpEndpoint)
 	}
-	if port == 0 {
-		port = 8545
+	if err := api.node.startHTTP(fmt.Sprintf("%s:%d", host, port), api.node.rpcAPIs, strings.Split(apis, ","), cors); err != nil {
+		return false, err
 	}
-
-	corsDomains := strings.Split(cors, " ")
-	err := rpc.StartHTTP(address, port, corsDomains, offeredAPIs)
-	return err == nil, err
+	return true, nil
 }
 
 // StopRPC terminates an already running HTTP RPC API endpoint.
 func (api *PrivateAdminAPI) StopRPC() (bool, error) {
-	err := rpc.StopHTTP()
-	return err == nil, err
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
+
+	if api.node.httpHandler == nil {
+		return false, fmt.Errorf("HTTP RPC not running")
+	}
+	api.node.stopHTTP()
+	return true, nil
 }
 
 // StartWS starts the websocket RPC API server.
-func (api *PrivateAdminAPI) StartWS(address string, port int, cors string, apis string) (bool, error) {
-	var offeredAPIs []rpc.API
-	if len(apis) > 0 {
-		namespaces := set.New()
-		for _, a := range strings.Split(apis, ",") {
-			namespaces.Add(strings.TrimSpace(a))
-		}
-		for _, api := range api.node.APIs() {
-			if namespaces.Has(api.Namespace) {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	} else {
-		// use by default all public API's
-		for _, api := range api.node.APIs() {
-			if api.Public {
-				offeredAPIs = append(offeredAPIs, api)
-			}
-		}
-	}
+func (api *PrivateAdminAPI) StartWS(host string, port int, cors string, apis string) (bool, error) {
+	api.node.lock.Lock()
+	defer api.node.lock.Unlock()
 
-	if address == "" {
-		address = "127.0.0.1"
+	if api.node.wsHandler != nil {
+		return false, fmt.Errorf("WebSocker RPC already running on %s", api.node.wsEndpoint)
 	}
-	if port == 0 {
-		port = 8546
+	if err := api.node.startWS(fmt.Sprintf("%s:%d", host, port), api.node.rpcAPIs, strings.Split(apis, ","), cors); err != nil {
+		return false, err
 	}
-
-	corsDomains := strings.Split(cors, " ")
-
-	err := rpc.StartWS(address, port, corsDomains, offeredAPIs)
-	return err == nil, err
+	return true, nil
 }
 
 // StopRPC terminates an already running websocket RPC API endpoint.
 func (api *PrivateAdminAPI) StopWS() (bool, error) {
-	err := rpc.StopWS()
-	return err == nil, err
+	api.node.lock.Lock()
```

Comment:
Instead of locking in the API it might be better to lock in the node (e.g. `node.stopWS`) since that is where the race conditions can occur (same for the HTTP interface). I would also move the check if the interface is running to the node.